File: embeddeds/java/src/main/java/io/github/mosser/arduinoml/embedded/java/dsl/InstructionBuilder.java
package io.github.mosser.arduinoml.embedded.java.dsl;


import io.github.mosser.arduinoml.kernel.behavioral.Action;
import io.github.mosser.arduinoml.kernel.behavioral.AnalogAction;
import io.github.mosser.arduinoml.kernel.behavioral.DigitalAction;
import io.github.mosser.arduinoml.kernel.behavioral.DigitalCondition;
import io.github.mosser.arduinoml.kernel.structural.Actuator;
import io.github.mosser.arduinoml.kernel.structural.Constant;
import io.github.mosser.arduinoml.kernel.structural.SIGNAL;

import java.util.Optional;

public class InstructionBuilder {

    private StateBuilder parent;

    private DigitalAction localDigital = new DigitalAction();
    private AnalogAction localAnalog = new AnalogAction();

    InstructionBuilder(StateBuilder parent, String target) {
        this.parent = parent;
        Optional<Actuator> opt = parent.parent.findActuator(target);
        Actuator act = opt.orElseThrow(() -> new IllegalArgumentException("Illegal actuator: ["+target+"]"));
        localDigital.setActuator(act);
        localAnalog.setActuator(act);
    }

    public StateBuilder toHigh() { localDigital.setValue(SIGNAL.HIGH); return goUpDigital(); }

    public StateBuilder toLow() { localDigital.setValue(SIGNAL.LOW); return goUpDigital(); }

    public StateBuilder toValue(float value){
        localAnalog.setValue(new Constant("constant", value)); // TODO! Might change with constant modifications in kernel
        return goUpAnalog();
    }

    private StateBuilder goUpDigital() {
        parent.local.getActions().add(this.localDigital);
        return parent;
    }

    private StateBuilder goUpAnalog() {
        parent.local.getActions().add(this.localAnalog);
        return parent;
    }

}

File: embeddeds/python/pyArduinoML/methodchaining/InstructionBuilder.py
from pyArduinoML.model.DigitalAction import DigitalAction
from pyArduinoML.model.SIGNAL import Signal


class InstructionBuilder:
    """
    Builder for defining instructions (actions) for actuators within a state.
    """

    def __init__(self, state_builder, actuator_name):
        """
        Constructor for InstructionBuilder.

        :param state_builder: StateBuilder, the parent builder.
        :param actuator_name: str, the name of the actuator.
        """
        self.state_builder = state_builder
        self.actuator_name = actuator_name
        self.signal = None

    def toHigh(self):
        """
        Set the actuator to HIGH in this instruction.

        :return: StateBuilder, the parent builder for chaining.
        """
        actuator = self.state_builder.parent.findActuator(self.actuator_name)
        if not actuator:
            raise ValueError(f"Actuator {self.actuator_name} not found.")
        action = DigitalAction(Signal.HIGH, actuator)
        self.state_builder.local.actions.append(action)
        return self.state_builder

    def toLow(self):
        """
        Set the actuator to LOW in this instruction.

        :return: StateBuilder, the parent builder for chaining.
        """
        actuator = self.state_builder.parent.findActuator(self.actuator_name)
        if not actuator:
            raise ValueError(f"Actuator {self.actuator_name} not found.")
        action = DigitalAction(Signal.LOW, actuator)
        self.state_builder.local.actions.append(action)
        return self.state_builder
